src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java

```diff
@@ -472,43 +481,176 @@ public void onClusterServiceClose() {
                 @Override
                 public void onTimeout(TimeValue timeout) {
                     // Try one more time...
-                    doStart();
+                    start();
                 }
             });
         }
 
-        void performOnPrimary(int primaryShardId, final ShardRouting shard) {
+        /** upon success, finish the first phase and transfer responsibility to the {@link ReplicationPhase} */
+        void finishAndMoveToReplication(ReplicationPhase replicationPhase) {
+            if (finished.compareAndSet(false, true)) {
+                replicationPhase.start();
+            } else {
+                assert false : "finishAndMoveToReplication called but operation is already finished";
+            }
+        }
+
+
+        void finishAsFailed(Throwable failure) {
```

Comment:
Is this supposed to be the hook for the counter dec? It seems to be called after remote execution failed (https://github.com/elastic/elasticsearch/pull/10749/files#diff-a1076062ccd18e89873d62fe1d78e1e1R453) sometimes when the operation could not be executed at all due to blocks (https://github.com/elastic/elasticsearch/pull/10749/files#diff-a1076062ccd18e89873d62fe1d78e1e1R363) and also if write locally on the primary failed (https://github.com/elastic/elasticsearch/pull/10749/files#diff-a1076062ccd18e89873d62fe1d78e1e1R408) I would still have to be as careful with where to increment and decrement the counter like before.

I was hoping that this refactoring separates remote and local execution for the primary operation more strictly because this is I think where the major difficulty with reasoning about what the code does stems from. I might not fully understand though...